### PR TITLE
Issue3395/dpl 111 as a user i would like the ability to display tube racks {heron} {tor 67}

### LINKED ISSUE #3395: DPL-111 - As a user I would like the ability to display tube racks [Heron] [TOR-67]
This represents the sequencescape work required for https://github.com/sanger/limber/issues/794

Original issue:
**User story**
As a user I would like the ability to display tube racks [Heron]

Require the addition of a few attributes to the API.

**Who are the primary contacts for this story**
Iraad / Steve / James

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] I would like to be able to view tube racks
- [ ] ...

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

References
This story has a non-blocking relationship with:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**
Users want to generate CSV containing the concentration of the source tubes. (I suggested that we could just generate the robot driver files, rather then them needing to copy-paste the information into an intermediate spreadsheet). However, both approaches require a means of grouping together multiple tubes for processing at the same time.

While this doesn’t need to be a tube rack, we:

- Already have the back-end support for this
- Need to add tube racks shortly for other purposes

A meeting about Tapestation for Stef's work led to a conversation about this work for Naomi as well.  See https://github.com/sanger/quanthub/issues/102#issuecomment-906259713

### PR DESCRIPTION
**Pull request by JamesGlover**
Closes #3395 

This adds some behind-the scenes work to enable tube rack viewing in limber.

- Tube racks have have and display comments
- Add a few more attributes and associations to the API
- Refactor some of the validation code
- Used stubbed factories for record testing to improve performance

### CODE REVIEW

File: spec/resources/api/v2/tube_resource_spec.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@
 RSpec.describe Api::V2::TubeResource, type: :resource do
   subject(:resource) { described_class.new(resource_model, {}) }
 
-  let(:resource_model) { create :tube, barcode_number: 1 }
+  let(:resource_model) { build_stubbed :tube, barcode_number: 1 }
```
Comment (sdjmchattie) [inline spec/resources/api/v2/tube_resource_spec.rb:9]: Heh .. well today I learned a new trick!

**Review by sdjmchattie — APPROVED** [review]

**Comment by sdjmchattie** [comment]
Worth noting, it turns out it's impossible to satisfy both the push linter and the pull request linter because the version of Prettier on `develop` is different from in this branch and they don't agree with each other.  I've currently got it set up to be happy in the destination `develop` branch.

**Comment by JamesGlover** [comment]
Ahhh, that's interesting. Suppose the trick for solving that kind of thing would be to rebase on HEAD.

**Comment by sdjmchattie** [comment]
> Ahhh, that's interesting. Suppose the trick for solving that kind of thing would be to rebase on HEAD.

Would have been good, except the code on develop didn't appear to satisfy its own prettier version so you'd have all that to deal with as well, perhaps.

**Pull request merged**